File: translate_tours.py
import json
import os
import urllib.parse
import urllib.request
import time
from concurrent.futures import ThreadPoolExecutor, as_completed

locales = ['es', 'fr', 'de', 'it', 'pt', 'ru', 'nl', 'zh']
cache_file = './database/seeders/data/translations_cache.json'
tours_file = './database/seeders/data/tours.json'

# Load cache
cache = {}
if os.path.exists(cache_file):
    try:
        with open(cache_file, 'r', encoding='utf-8') as f:
            cache = json.load(f)
    except Exception as e:
        print(f"Error loading cache: {e}")
        cache = {}

print(f"Loaded {len(cache)} unique English strings from cache.")

# Load tours
if not os.path.exists(tours_file):
    print("Error: tours.json not found!")
    exit(1)

with open(tours_file, 'r', encoding='utf-8') as f:
    tours = json.load(f)

# Collect all unique strings needing translation
strings = set()

def collect(text):
    if not text:
        return
    text = text.strip()
    if text.isdigit():
        return
    strings.add(text)

for tour in tours:
    collect(tour.get('title', ''))
    collect(tour.get('description', ''))
    collect(tour.get('duration', ''))
    collect(tour.get('nights', ''))
    collect(tour.get('startingPoint', ''))
    collect(tour.get('arrivalCity', ''))
    collect(tour.get('accommodation', ''))
    collect(tour.get('guide', ''))
    collect(tour.get('tripType', ''))
    collect(tour.get('difficulty', ''))
    
    # Parse comma-separated languages
    if tour.get('languages'):
        langs = [l.strip() for l in tour['languages'].split(',')]
        for lang in langs:
            collect(lang)
            
    if tour.get('included'):
        for item in tour['included']:
            collect(item)
            
    if tour.get('excluded'):
        for item in tour['excluded']:
            collect(item)
            
    if tour.get('itinerary'):
        for item in tour['itinerary']:
            collect(item.get('day', ''))
            collect(item.get('title', ''))
            collect(item.get('description', ''))

unique_strings = list(strings)
print(f"Total unique English strings to translate: {len(unique_strings)}")

# Google Translate function using urllib
def translate_api(text, target, source='en'):
    if not text.strip():
        return ""
    
    url = "https://translate.googleapis.com/translate_a/single?client=gtx&sl=" + source + "&tl=" + target + "&dt=t&q=" + urllib.parse.quote(text)
    
    req = urllib.request.Request(
        url,
        headers={'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'}
    )
    
    try:
        with urllib.request.urlopen(req, timeout=10) as response:
            if response.status != 200:
                return None
            data = json.loads(response.read().decode('utf-8'))
            if not data or not data[0]:
                return None
            translated_text = ""
            for line in data[0]:
                if line and line[0]:
                    translated_text += line[0]
            return translated_text
    except Exception as e:
        return None

# Build task list
tasks = []
for str_val in unique_strings:
    if str_val not in cache:
        cache[str_val] = {}
    for locale in locales:
        if locale not in cache[str_val]:
            tasks.append((str_val, locale))

print(f"Remaining translation operations: {len(tasks)}")

if not tasks:
    print("All translations are up to date!")
    exit(0)

# We will save cache in chunks to be thread-safe
completed_count = 0
failed_count = 0

def save_cache():
    with open(cache_file, 'w', encoding='utf-8') as f:
        json.dump(cache, f, ensure_ascii=False, indent=2)

def worker(task):
    str_val, locale = task
    result = translate_api(str_val, locale)
    return str_val, locale, result

# Use ThreadPoolExecutor to run tasks in parallel
max_threads = 15
print(f"Starting translations using ThreadPoolExecutor with {max_threads} threads...")

with ThreadPoolExecutor(max_workers=max_threads) as executor:
    futures = {executor.submit(worker, task): task for task in tasks}
    
    for future in as_completed(futures):
        str_val, locale, result = future.result()
        if result is not None:
            cache[str_val][locale] = result
            completed_count += 1
            if completed_count % 50 == 0:
                save_cache()
                print(f"Progress: {completed_count} translations saved.")
        else:
            failed_count += 1
            # print(f"Failed: {str_val[:30]}... to {locale}")

save_cache()
print(f"\nCompleted {completed_count} translations. Failed: {failed_count}.")

File: translate_tours.php
<?php

/**
 * Script to translate all tour data in tours.json into the 8 supported locales
 * using the Google Translate API, storing them in a translation cache.
 */

$locales = ['es', 'fr', 'de', 'it', 'pt', 'ru', 'nl', 'zh'];
$cacheFile = __DIR__ . '/database/seeders/data/translations_cache.json';

// Load cache
$cache = [];
if (file_exists($cacheFile)) {
    $cache = json_decode(file_get_contents($cacheFile), true);
    if (!is_array($cache)) {
        $cache = [];
    }
}
echo "Loaded " . count($cache) . " unique English strings from cache.\n";

// Load tours
$toursFile = __DIR__ . '/database/seeders/data/tours.json';
if (!file_exists($toursFile)) {
    die("Error: tours.json not found!\n");
}
$tours = json_decode(file_get_contents($toursFile), true);

// Collect all unique strings needing translation
$strings = [];

function collect($text) {
    global $strings;
    if ($text === null || $text === '') return;
    $text = trim($text);
    if (is_numeric($text)) return;
    $strings[$text] = true;
}

foreach ($tours as $tour) {
    collect($tour['title'] ?? '');
    collect($tour['description'] ?? '');
    collect($tour['duration'] ?? '');
    collect($tour['nights'] ?? '');
    collect($tour['startingPoint'] ?? '');
    collect($tour['arrivalCity'] ?? '');
    collect($tour['accommodation'] ?? '');
    collect($tour['guide'] ?? '');
    collect($tour['tripType'] ?? '');
    collect($tour['difficulty'] ?? '');
    
    // Parse comma-separated languages
    if (!empty($tour['languages'])) {
        $langs = array_map('trim', explode(',', $tour['languages']));
        foreach ($langs as $lang) {
            collect($lang);
        }
    }
    
    if (!empty($tour['included'])) {
        foreach ($tour['included'] as $item) {
            collect($item);
        }
    }
    if (!empty($tour['excluded'])) {
        foreach ($tour['excluded'] as $item) {
            collect($item);
        }
    }
    if (!empty($tour['itinerary'])) {
        foreach ($tour['itinerary'] as $item) {
            collect($item['day'] ?? '');
            collect($item['title'] ?? '');
            collect($item['description'] ?? '');
        }
    }
}

$uniqueStrings = array_keys($strings);
echo "Total unique English strings to translate: " . count($uniqueStrings) . "\n";

// Google Translate function
function translate($text, $target, $source = 'en') {
    if (trim($text) === '') return '';
    
    $url = "https://translate.googleapis.com/translate_a/single?client=gtx&sl=" . $source . "&tl=" . $target . "&dt=t&q=" . urlencode($text);
    
    $ch = curl_init();
    curl_setopt($ch, CURLOPT_URL, $url);
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
    curl_setopt($ch, CURLOPT_USERAGENT, 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36');
    curl_setopt($ch, CURLOPT_TIMEOUT, 10);
    
    $response = curl_exec($ch);
    $httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);
    curl_close($ch);
    
    if ($httpCode !== 200 || !$response) {
        return null;
    }
    
    $result = json_decode($response, true);
    if (!isset($result[0]) || !is_array($result[0])) {
        return null;
    }
    
    $translatedText = "";
    foreach ($result[0] as $line) {
        if (isset($line[0])) {
            $translatedText .= $line[0];
        }
    }
    return $translatedText;
}

$totalToTranslate = 0;
foreach ($uniqueStrings as $str) {
    foreach ($locales as $locale) {
        if (!isset($cache[$str][$locale])) {
            $totalToTranslate++;
        }
    }
}

echo "Remaining translation operations: {$totalToTranslate}\n";
if ($totalToTranslate === 0) {
    echo "All translations are up to date!\n";
    exit(0);
}

$count = 0;
$skipped = 0;
$failed = 0;

foreach ($uniqueStrings as $str) {
    if (!isset($cache[$str])) {
        $cache[$str] = [];
    }
    
    foreach ($locales as $locale) {
        if (isset($cache[$str][$locale])) {
            continue;
        }
        
        echo "Translating to [{$locale}]: \"" . mb_substr($str, 0, 50) . "...\"\n";
        
        $translated = translate($str, $locale);
        
        if ($translated !== null) {
            $cache[$str][$locale] = $translated;
            $count++;
            
            // Save cache every 15 translation calls
            if ($count % 15 === 0) {
                file_put_contents($cacheFile, json_encode($cache, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES));
                echo "Saved progress: {$count} translations completed.\n";
            }
            
            // Wait 50ms to be gentle with the API
            usleep(50000);
        } else {
            $failed++;
            echo "FAILED to translate to [{$locale}]: \"" . mb_substr($str, 0, 50) . "\"\n";
            // If we hit too many failures, pause or exit to prevent spamming
            if ($failed > 10) {
                file_put_contents($cacheFile, json_encode($cache, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES));
                die("Too many API failures. Stopping to avoid rate limits. Please try running the script again.\n");
            }
            // Sleep longer on failure
            sleep(2);
        }
    }
}

// Final save
file_put_contents($cacheFile, json_encode($cache, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES));
echo "\nDone! Successfully translated {$count} strings. Failed: {$failed}.\n";
